python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -0,0 +1,227 @@
+import logging
+from types import ModuleType
+from typing import TYPE_CHECKING, List, Optional
+
+import ray
+import ray.util.serialization
+from ray.experimental.channel import ChannelContext
+from ray.experimental.channel.common import ChannelInterface
+from ray.experimental.channel.nccl_group import _NcclGroup
+from ray.util.annotations import DeveloperAPI
+
+if TYPE_CHECKING:
+    import torch
+
+    from ray.experimental.channel.torch_tensor_type import (
+        TorchTensorType,
+        _TorchTensorWrapper,
+    )
+
+
+# Logger for this module. It should be configured at the entry point
+# into the program using Ray. Ray provides a default configuration at
+# entry/init points.
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class TorchTensorNcclChannel(ChannelInterface):
+    def __init__(
+        self,
+        writer: ray.actor.ActorHandle,
+        readers: List[ray.actor.ActorHandle],
+        typ: "TorchTensorType",
+        device: Optional["torch.device"] = None,
+    ):
+        import torch
+
+        from ray.air._internal import torch_utils
+
+        self.torch: ModuleType = torch
+
+        self._nccl_group = None
+
+        self._writer = writer
+        self._writer_rank: Optional[int] = None
+        self._readers = readers
+        self._reader_ranks: Optional[List[int]] = None
+        self._writer_registered: bool = False
+        self._reader_registered: bool = False
+
+        self._device = device
+        if self._device is None:
+            # TODO(swang): Allow default device to be overridden.
+            self._device = torch_utils.get_devices()[0]
+        if self._device.type != "cuda":
+            raise ValueError(
+                f'Actor\'s default device has type "{self._device.type}", need "cuda"'
+            )
+
+        assert typ.transport == "nccl"
+        self._typ = typ
+
+        ctx = ChannelContext.get_current()
+        self._nccl_group = ctx.nccl_group
+        assert (
+            self._nccl_group is not None
+        ), "ChannelContext.nccl_group is not initialized."
+
+        self._writer_rank = self._nccl_group.get_rank(self._writer)
+        self._reader_ranks = [
+            self._nccl_group.get_rank(reader) for reader in self._readers
+        ]
+
+        if (
+            self._writer_rank is not None
+            and self._writer_rank == self._nccl_group.get_self_rank()
+        ):
+            self._writer_registered = True
+
+        if (
+            self._reader_ranks
+            and self._nccl_group.get_self_rank() in self._reader_ranks
+        ):
+            self._reader_registered = True
+
+    def ensure_registered_as_writer(self):
+        assert self._nccl_group is not None, "Actor is not part of a NCCL group"
+        return self._writer_registered, "Actor is not the writer"
```

Comment:
Is it intended to return a tuple here and below?

Reply:
Ah, thanks, this was a typo!